File: January 23/24 Jan/909. Snakes and Ladders LC.cpp
class Solution {
public:
    pair<int,int> helper(int n, int ind)
    {
        ind--;
        int _i = (n-1) - (ind/n);
        int _j = ind%n;
        if((ind/n)&1) _j = (n-1) - (ind%n);

        return {_i,_j};
    }
    int snakesAndLadders(vector<vector<int>>& board) {
        int n=board.size();
        
        queue<int> q;
        q.push(1);

        int qs, ind, step=0;
        pair<int,int> index = helper(n,1);
        board[index.first][index.second]=-2;
        while(!q.empty())
        {
            qs=q.size();
            for(int c=0;c<qs;c++)
            {
                ind = q.front();
                q.pop();

                if(ind==(n*n)) return step;

                for(int i=ind+1;i<=min(ind+6,n*n);i++)
                {
                    index = helper(n,i);
                    if(board[index.first][index.second] == -1)
                    {
                        q.push(i);
                        board[index.first][index.second]=-2;
                    }
                    else if(board[index.first][index.second] > 0)
                    {
                        q.push(board[index.first][index.second]);
                        board[index.first][index.second]=-2;
                    }
                }
            }
            step++;
        }

        return -1;
    }
};

File: MAY 2025/31 May/909. Snakes and Ladders LC.java
class Solution {
    class Pair{
        int NoOfDiceRoll, cell;
        public Pair(int cell, int roll)
        {
            this.cell = cell;
            this.NoOfDiceRoll = roll;
        }
    }
    public int snakesAndLadders(int[][] board) {
        int N = board.length;
        int[] flatBoard = new int[N*N+5];
        Arrays.fill(flatBoard, N*N+1);

        for(int row=0; row<N; row++)
        {
            for(int col=0; col<N; col++)
            {
                int cell = row*N+col+1;
                
                if(row%2==0)
                    flatBoard[cell] = board[N-1-row][col];
                else
                    flatBoard[cell] = board[N-1-row][N-1-col];
                // System.out.println(row+" "+col+" : "+cell+" flatboard = "+flatBoard[cell]);
            }

        }
        
        Queue<Pair> q = new LinkedList<>();
        q.offer(new Pair(1,0));
        boolean[] visited = new boolean[N*N+5];
        visited[1] = true;

        while(!q.isEmpty())
        {
            Pair pr = q.poll();
            int distance = pr.NoOfDiceRoll, cell = pr.cell;

            if(cell==(N*N))
                return distance;
            
            for(int next=cell+1; next<=Math.min(N*N, cell+6); next++)
            {
                int dest = (flatBoard[next] == -1) ? next : flatBoard[next];
                if (!visited[dest]) {
                    visited[dest] = true;
                    q.offer(new Pair(dest, distance+1));
                }          
            }
        }

        return -1;
    }
}
